xarray_beam/_src/core.py

```diff
@@ -370,26 +410,33 @@ def _shard_inputs(self) -> List[Tuple[Optional[int], Optional[str]]]:
       return [(i, None) for i in range(self.shard_count)]
 
     inputs = []
-    for name, variable in self.dataset.items():
+    for name, variable in self._first.items():
       if self.sharded_dim in variable.dims:
         inputs.extend([(i, name) for i in range(self.shard_count)])
       else:
         inputs.append((None, name))
     return inputs
 
-  def _key_to_chunks(self, key: Key) -> Iterator[Tuple[Key, xarray.Dataset]]:
+  def _key_to_chunks(self, key: Key) -> Iterator[Tuple[Key, DatasetOrDatasets]]:
     """Convert a Key into an in-memory (Key, xarray.Dataset) pair."""
     sizes = {
         dim: self.expanded_chunks[dim][self.offset_index[dim][offset]]
         for dim, offset in key.offsets.items()
     }
     slices = offsets_to_slices(key.offsets, sizes)
-    dataset = self.dataset if key.vars is None else self.dataset[list(key.vars)]
-    chunk = dataset.isel(slices)
-    # Load the data, using a separate thread for each variable
-    num_threads = len(self.dataset)
-    result = chunk.chunk().compute(num_workers=num_threads)
-    yield key, result
+    results = []
+    for ds in self._datasets:
+      dataset = ds if key.vars is None else ds[list(key.vars)]
+      chunk = dataset.isel(slices)
+      # Load the data, using a separate thread for each variable
+      num_threads = len(dataset)
+      result = chunk.chunk().compute(num_workers=num_threads)
+      results.append(result)
+
+    if type(self.dataset) is xarray.Dataset:
+      yield key, results[0]
+    else:
+      yield key, tuple(results)
```

Comment:
To be consistent with the type annotation:
```suggestion
      yield key, list(results)
```